Server/Application/Application.m

```diff
@@ -32,22 +34,60 @@ + (XCUIApplication *)currentApplication {
 - (void)startSession {
     DDLogDebug(@"Launching application '%@'", self.app.bundleID);
 
+    // In some contexts, the application has not been completely installed
+    // when the POST /session route is called. In that case, the launch will
+    // fail with a detectable error.
+    //
+    // The private LSApplicationWorkspace API does not work on physical devices
+    // so we cannot poll to wait for the application to install.
+    NSUInteger attempts = 1;
+    NSUInteger maxAttempts = 12;
+    NSTimeInterval sleepBetween = 5;
+    NSTimeInterval start = [[CBXMachClock sharedClock] absoluteTime];
+
     __block NSError *outerError = nil;
-    [ThreadUtils runSync:^(BOOL *setToTrueWhenDone) {
-        [[Testmanagerd get] _XCT_launchApplicationWithBundleID:self.app.bundleID
-                                                     arguments:self.app.launchArguments
-                                                   environment:self.app.launchEnvironment
-                                                    completion:^(NSError *innerError) {
-                                                        outerError = innerError;
-                                                        *setToTrueWhenDone = YES;
-                                                    }];
-    }];
-
-    if (outerError) {
+
+    do {
+        [ThreadUtils runSync:^(BOOL *setToTrueWhenDone) {
+            [[Testmanagerd get]
+             _XCT_launchApplicationWithBundleID:self.app.bundleID
+             arguments:self.app.launchArguments
+             environment:self.app.launchEnvironment
+             completion:^(NSError *innerError) {
+                outerError = innerError;
+                *setToTrueWhenDone = YES;
+            }];
+        }];
+
+        if (!outerError) {
+            break;
+        }
+
         NSString *errorMessage;
-        errorMessage = [NSString stringWithFormat:@"Could not launch application with bundle identifier: %@\n%@",
-                        self.app.bundleID, outerError.localizedDescription];
+        errorMessage = [NSString stringWithFormat:@"Attempt %@ of %@ - could not "
+                        "launch application with bundle identifier: %@\n%@",
+                        @(attempts), @(maxAttempts), self.app.bundleID,
+                        outerError.localizedDescription];
+
+        DDLogDebug(@"Application did not launch (in the loop): Application is installed?  %@",
+                   @([CBLSApplicationWorkspace applicationIsInstalled:self.app.bundleID]));
+
+
+        CFRunLoopRunInMode(kCFRunLoopDefaultMode, sleepBetween, false);
+
+    } while (attempts < maxAttempts + 1);
+
+    NSTimeInterval elapsed = [[CBXMachClock sharedClock] absoluteTime] - start;
+
+    if (outerError && attempts == maxAttempts) {
```

Comment:
If we get here and outerError != null then I think we should throw the exception. attempts == maxAttempts should be true, but by adding it to the if we risk the code not working if we have an error in the handling of attempts

Reply:
Yes.  I agree.